controller/search_blackbox_test.go

```diff
@@ -962,3 +970,93 @@ func (s *searchBlackBoxTest) TestSearchQueryScenarioDriven() {
+// wi3 is child of wi2
+func (s *searchBlackBoxTest) TestIncludedParents() {
+	spaceOwner, err := testsupport.CreateTestIdentity(s.DB, testsupport.CreateRandomValidTestName("TestSearchQueryScenarioDriven-"), "TestWISearch")
+	require.Nil(s.T(), err)
+	spaceInstance := CreateSecuredSpace(s.T(), gormapplication.NewGormDB(s.DB), s.Configuration, *spaceOwner)
+	spaceIDStr := spaceInstance.ID.String()
+	wirepo := workitem.NewWorkItemRepository(s.DB)
+	workItemLinkCtrl := NewWorkItemLinkController(s.svc, gormapplication.NewGormDB(s.DB), s.Configuration)
+	require.NotNil(s.T(), workItemLinkCtrl)
+
+	parentWI0, err := wirepo.Create(
+		s.ctx, *spaceInstance.ID, workitem.SystemFeature,
+		map[string]interface{}{
+			workitem.SystemTitle: "Parent WI",
+			workitem.SystemState: workitem.SystemStateResolved,
+		}, spaceOwner.ID)
+	require.Nil(s.T(), err)
+
+	parentWI1, err := wirepo.Create(
+		s.ctx, *spaceInstance.ID, workitem.SystemTask,
+		map[string]interface{}{
+			workitem.SystemTitle: "Parent WI",
+			workitem.SystemState: workitem.SystemStateResolved,
+		}, spaceOwner.ID)
+	require.Nil(s.T(), err)
+
+	childWI, err := wirepo.Create(
+		s.ctx, *spaceInstance.ID, workitem.SystemBug,
+		map[string]interface{}{
+			workitem.SystemTitle: "Child WI",
+			workitem.SystemState: workitem.SystemStateResolved,
+		}, spaceOwner.ID)
+	require.Nil(s.T(), err)
+	// create parent links
+	createPayload := newCreateWorkItemLinkPayload(parentWI0.ID, parentWI1.ID, link.SystemWorkItemLinkTypeParentChildID)
+	_, workItemLink := test.CreateWorkItemLinkCreated(s.T(), s.svc.Context, s.svc, workItemLinkCtrl, createPayload)
+	require.NotNil(s.T(), workItemLink)
+	require.Equal(s.T(), parentWI0.ID, workItemLink.Data.Relationships.Source.Data.ID)
+	require.Equal(s.T(), parentWI1.ID, workItemLink.Data.Relationships.Target.Data.ID)
+	require.Equal(s.T(), link.SystemWorkItemLinkTypeParentChildID, workItemLink.Data.Relationships.LinkType.Data.ID)
+
+	createPayload = newCreateWorkItemLinkPayload(parentWI1.ID, childWI.ID, link.SystemWorkItemLinkTypeParentChildID)
+	_, workItemLink = test.CreateWorkItemLinkCreated(s.T(), s.svc.Context, s.svc, workItemLinkCtrl, createPayload)
+	require.NotNil(s.T(), workItemLink)
+	require.Equal(s.T(), parentWI1.ID, workItemLink.Data.Relationships.Source.Data.ID)
+	require.Equal(s.T(), childWI.ID, workItemLink.Data.Relationships.Target.Data.ID)
+	require.Equal(s.T(), link.SystemWorkItemLinkTypeParentChildID, workItemLink.Data.Relationships.LinkType.Data.ID)
+
+	filter := fmt.Sprintf(`{"$AND": [{"space": "%s"}]}`, spaceIDStr)
+	_, result := test.ShowSearchOK(s.T(), nil, nil, s.controller, &filter, nil, nil, nil, nil, &spaceIDStr)
+	require.NotEmpty(s.T(), result.Data)
+	require.Len(s.T(), result.Data, 3)
+	require.Len(s.T(), result.Included, 2)
+
+	// verify included objects
+	includedMustHave := []string{parentWI0.ID.String(), parentWI1.ID.String()}
+	for _, ele := range result.Included {
+		if appWI, ok := ele.(app.WorkItem); ok == true {
```

Comment:
This would reduce the "depth" or nesting of this loop and improve the readability IMHO. Notice that empty structs (`struct{}`) don't consume any memory and that it is safe to delete a an entry from a map by a key that is not present in the map.

```go
includedMustHave := map[uuid.UUID]struct{}{
	parentWI0.ID: struct{}{},
	parentWI1.ID: struct{}{},
}
for _, ele := range result.Included {
	appWI, ok := ele.(app.WorkItem)
	if !ok || appWI.Type != APIStringTypeWorkItem {
		continue
	}
	delete(includedMustHave, appWI.ID)
}
assert.Empty(s.T(), includedMustHave)
```

Reply:
Sure, done in https://github.com/fabric8-services/fabric8-wit/pull/1592/commits/d64aa3a7bd28e06ea6a169926ea1e1f3b09216b0
and it looks better, thanks!